server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -10,48 +10,102 @@
 
 import org.apache.lucene.index.LeafReaderContext;
 import org.elasticsearch.common.document.DocumentField;
-import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.mapper.IgnoredFieldMapper;
+import org.elasticsearch.index.mapper.LegacyTypeFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.mapper.RoutingFieldMapper;
+import org.elasticsearch.index.mapper.SourceFieldMapper;
+import org.elasticsearch.index.query.SearchExecutionContext;
 import org.elasticsearch.search.fetch.FetchContext;
 import org.elasticsearch.search.fetch.FetchSubPhase;
 import org.elasticsearch.search.fetch.FetchSubPhaseProcessor;
+import org.elasticsearch.search.fetch.StoredFieldsContext;
 import org.elasticsearch.search.fetch.StoredFieldsSpec;
 
 import java.io.IOException;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
 
 /**
  * A fetch sub-phase for high-level field retrieval. Given a list of fields, it
  * retrieves the field values through the relevant {@link org.elasticsearch.index.mapper.ValueFetcher}
  * and returns them as document fields.
  */
 public final class FetchFieldsPhase implements FetchSubPhase {
+
+    private static final List<FieldAndFormat> DEFAULT_METADATA_FIELDS = List.of(
+        new FieldAndFormat(IgnoredFieldMapper.NAME, null),
+        new FieldAndFormat(RoutingFieldMapper.NAME, null),
+        new FieldAndFormat(LegacyTypeFieldMapper.NAME, null)
+    );
+
     @Override
     public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
-        FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
-        if (fetchFieldsContext == null) {
+        final FetchFieldsContext fetchFieldsContext = fetchContext.fetchFieldsContext();
+        final StoredFieldsContext storedFieldsContext = fetchContext.storedFieldsContext();
+
+        boolean fetchStoredFields = storedFieldsContext != null && storedFieldsContext.fetchFields();
+        if (fetchFieldsContext == null && fetchStoredFields == false) {
             return null;
         }
 
-        FieldFetcher fieldFetcher = FieldFetcher.create(fetchContext.getSearchExecutionContext(), fetchFieldsContext.fields());
+        final SearchExecutionContext searchExecutionContext = fetchContext.getSearchExecutionContext();
+        final FieldFetcher fieldFetcher = fetchFieldsContext == null ? null
+            : fetchFieldsContext.fields() == null ? null
+            : fetchFieldsContext.fields().isEmpty() ? null
+            : FieldFetcher.create(searchExecutionContext, fetchFieldsContext.fields());
 
+        final FieldFetcher metadataFieldFetcher;
+        if (storedFieldsContext != null
+            && storedFieldsContext.fieldNames() != null
+            && storedFieldsContext.fieldNames().isEmpty() == false) {
```

Comment:
You still should check `storedFieldsContext.fetchFields()` here, or you are doing to end up fetching default metadata fields when `stored_fields: _none_` is requested (and fields are requested at the same time), which I think breaks backwards compatibility. Also, it would be inconsistent behaviour as `stored_fields: _none_` behaves differently depending on whether `fields` are requested or not.

Once you do that, you will need to add null checks around metadata field fetcher down below, like you do for the other fetcher already.

Reply:
Ok so I understand that here there is a missing test which is one where `stored_fields: _none_` and `fields` requests a few fields.

Reply:
I tested this on `main` before testing it on my branch

```
---
fetch fields with none stored_fields:
  - do:
      search:
        index: test
        body:
          stored_fields: _none_
          fields: [stored_keyword, keyword, stored_value, value, ignored_keyword, ignored_value, _ignored]

  - match: { hits.total.value: 1 }
```

Adding this to `520_fetch_fields.yaml` and in **main** I get the following validation error:

Reply:
I wrote the following **missing** test in **main** not in this branch:

```
---
fetch fields with none stored_fields:
  - do:
      search:
        index: test
        body:
          stored_fields: _none_
          fields: [stored_keyword, keyword, stored_value, value, ignored_keyword, ignored_value, _ignored]

  - match: { hits.total.value: 1 }
```

I see it failing with the following error:
```
[{"error":{"root_cause":[{"type":"action_request_validation_exception","reason":"Validation Failed: 1: [stored_fields] cannot be disabled when using the [fields] option;","stack_trace":"org.elasticsearch.ElasticsearchException$1: Validation Failed: 1: [stored_fields] cannot be disabled when using the [fields] option;
```

Is there anything wrong in my test?